README.md
# NNbenchmark

NNbenchmark was created during the Google Summer of Code, 2019 as a part of The R Project for Statistical Computing, to verify the convergence of the training algorithms provided
in 69 Neural Network R packages available on CRAN to date. Neural networks must be trained with second order algorithms and not with first order algorithms as many packages seem to do. 


The purpose of this project is to verify the quality of the training algorithms in R packages that provide neural network of perceptron type (one input layer, one normalized layer, one hidden layer with nonlinear activation function usually tanh(), 
one normalized layer, one output output layer) for regression purpose i.e. NN(X1, ..., Xn) = E[Y].



# Packages Tested  

This GSoC project will conduct a comprehensive survey of all packages that have the “neural network” keyword in thepackage title or in the package description. There are currently 69 packages on CRAN with this keyword.  


|   Packages    |               |               |               |               |               |
|:-------------:|:-------------:|:-------------:|:-------------:|:-------------:|:-------------:| 
|   AMORE       |    ANN2       |  appnn        | autoencoder   | automl        | BNN           |
|   brnn        |    Buddle     |  CaDENCE      |   cld2        | cld3          | condmixt      |
|   DALEX2      |    DamiaNN    |  deepnet      |   deepNN      | DNMF          | elmNNrcpp     |
|   ELMR        | EnsembleBase  |  evclass      | gamlss.add    | gcForest      | GMDH          |
|   GMDH2       |    GMDHreg    |  grnn         |   h2o         | hybridEnsemble|  isingLenzMC  |
|   keras       | kerasformula  |  kerasR       |   leabRa      | learNN        |  LilRhino     |
|   monmlp      | neural        |  neuralnet    |NeuralNetTools | NlinTS        |  nnet         |
| nnetpredint   | nnfor         |  onnx         |OptimClassifier|  OSTSC        |  pnn          |
|   polyreg     | predictoR     |  qrnn         |   QuantumOps  | quarrint      | radiant.model |
|   rasclass    | rcane         |  rminer       |   rmn         | RSNNS         |  ruta         |
| simpleNeural  | snnr          |  softmaxreg   | Sojourn.Data  | spnn          |  TeachNet     |
|  tensorflow   | tfestimators  |  trackdem     | TrafficBDE    | validann      |               |



# Evaluation Criteria
***
The algorithms were tested on 12 regression datasets (both univariate and multivariate) of varying complexity.  

The score for each package was based on the following parameters:  

* Documentation (0-3 stars or a binary value of 0/1)
* UtilFunction (0-3 stars or a binary value of 0/1)
* ConvergenceQuality (0-3 stars based on percentile method)
* ConvergenceTime (0-3 stars based on percentile method)


To obtain the final rating, we take a weighted average of these 4 columns (giving more weight to ConvergenceQuality and ConvergenceTime).



# Authors  

## Selected Students:

- Akshaj Verma
- Salsabila Mahdi

## Mentors:

- Patrice Kiener  
- Christophe Dutang
- John C. Nash


NEWS
= Package NNbenchmark =
- Akshaj Verma, Salabila Mahdi, Patrice Kiener
- Christophe Dutang, John Nash 



NNbenchmark2019 2.4.0 - 2020-06-21
==================================
- New RStudio project
- Moved to https://github.com/pkR-pkR/NNbenchmark2019
- Roxygen 7.1.0 with old/classical presentation
- a_intro.R: ds <- grep("^[m,u]", ls("package:NNbenchmark2019"), value = TRUE); ds
- replaces   ds <- grep("^[m,u]", ls("package:NNbenchmark"), value = TRUE); ds
- clearNN(): comment the 2 lines that clear .GlobalEnv. 
- clearNN(): add NNbenchmark2019 in the exclusion



NNbenchmark 2.3.7 - 2020-04-10
==============================
- ht() and ht9() recognize data.table

NNbenchmark 2.3.6 - 2020-01-11
==============================
- ht() and ht9() fully rewritten. Accept now array and data.frame with matrices inside
- README.md added

NNbenchmark 2.3.5 - 2019-08-22
==============================
- DESCRIPTION and a_intro.R: add the html links to GitHub and AkshajVerma.com

NNbenchmark 2.3.4 - 2019-08-21 - sent to CRAN
=============================================
- DESCRIPTION: Modify the title to match title case (recommanded by CRAN)

NNbenchmark 2.3.3 - 2019-08-21
==============================
- DESCRIPTION: Modify the title 
- All files: Add value to Rd files (return in roxygen comments) for every functions

NNbenchmark 2.3.2 - 2019-08-20
==============================
- prepareZZ: extend donttest to run examples in less than 5 seconds

NNbenchmark 2.3.1 - 2019-08-20
==============================
- add2csv: examples rewritten with tempfile()
- datalist rewritten with format (file_without_ext: obj1, ..., objn)
- use for this purpose the hidden function tools:::list_data_in_pkg()

NNbenchmark 2.3 - 2019-08-19
============================
- NNdatasets: Change uNeuroOne from 3 to 2 neurons and 10 to 7 parameters
- NNdatasets: mRef153: correct a typo in the number of neurons (from 4 to 3)
- u+mNNdatasets: Update the documentation with the number of parameters for each dataset
- prepareZZ: rearrange the examples with donttest to run less than 5 seconds
- DESCRIPTION: Yifu Yan as author of the timeR function + class (timeR package)
- Test CRAN check and prepare the upload to CRAN

NNbenchmark 2.2 - 2019-08-13
============================
- NNdataSummary()

NNbenchmark 2.1.1 - 2019-08-13
==============================
- clearNN() donotdetach now includes kableExtra, dplyr, stringr

NNbenchmark 2.1 - 2019-08-12
============================
- detachNN() (to move library(pkg) out of the loop in NNTemplate)

NNbenchmark 2.0.1 - 2019-08-04
==============================
- Date 2019-08-04 added at the first line of *.R files
- Linked to GitHub. Package is rebuilt before commit (to force change in NAMESPACE)

NNbenchmark 2.0 - 2019-08-04
============================
- column 'params' added to the timer table. Columns in the data.frame are:
- timer$stop(event, start, end, duration, RMSE, MAE, stars, params, comment)
- Connect to GitHub

NNbenchmark 1.9 - 2019-08-03
============================
- New disk
- timer$stop(event, RMSE, MAE, stars, comment, printmsg) column 'stars' added to the table

NNbenchmark 1.8.6 - 2019-07-31
==============================
- timer$stop(printmsg = TRUE) Use printmsg = FALSE to avoid printing

NNbenchmark 1.8.2 - 2019-07-30
==============================
- clearNN(): ZZ is (now) detached from the environments
- clearNN(): donotdetach now indludes packages rmarkdown, knitr and captioner
- timer$stop suppress the message on the time elapsed

NNbenchmark 1.8.1 - 2019-07-29
==============================
- add2csv(): create a new entry in the documentation different than clearNN (minor bug)

NNbenchmark 1.8 - 2019-07-28
============================
- NNdatasets now includes for each dataset the corresponding fmlaNN and nparNN
- prepareZZ(Zsm = ...) adds has a new item to the final list:
- Zxy is a scaled or not scaled data.frame or matrix, depending the zmd and scale parameters

NNbenchmark 1.7.2 - 2019-07-25
==============================
- DESCRIPTION Imports: pkgload
- clearNN() => uses pkgload::unload() to unload poperly S4 classes and dll

NNbenchmark 1.7.1 - 2019-07-25
==============================
- clearNN() => "tools:rstudio" added to the default packages (NULL) not be detached

NNbenchmark 1.7 - 2019-07-23
============================
- add2csv()
- funRMSE(), funMAE()
- MAE (Mean absolute Error) added to timer_class and data.frame
- clearNN() => tryCatch(...) added

NNbenchmark 1.6 - 2019-07-21
============================
- clearNN()

NNbenchmark 1.5 - 2019-06-30
============================
- ht(..., l = 3)
- ht9()
- NNdatasets
- prepareZZ(..., scale =) replaces prepareZZ(..., normalize) as 'scale' is the usual word
- for (mean, standard deviation) adjustment in R. 
- prepareZZ() now returns the formula y ~ x or y ~ x1 + ... + xn for any number of columns

NNbenchmark 1.4.1 - 2019-06-24
==============================
- Function plotNN(, TF) + lipoNN(, TF) replaces if (TRUE/FALSE) plot
- Examples prepareZZ() rewritten accordingly

NNbenchmark 1.4 - 2019-06-23
============================
- Function plotNN() + lipoNN()
- Examples prepareZZ() rewritten accordingly

NNbenchmark 1.3 - 2019-06-23
============================
- Function prepareZZ()
- timeR: add a column RMSE 

NNbenchmark 1.2 - 2019-06-22
============================
- Function prepareZ()
- DESCRIPTION Suggests: brnn, validann

NNbenchmark 1.1 - 2019-06-17
============================
- Functions and classes timeR

NNbenchmark 1.0 - 2019-06-13
============================
- Functions cc() and ht()
- Datasets are now uNNdatasets.rda and mNNdatasets.rda 'dataNNbenchmark'
- LazyData: true' in DESCRIPTION file




R/ht.R
## ht 2019-08-04. Rewritten in 2020-11-01


## ht -> Voir package matsindf: Matrices in data frames


#' @title Concatenates head() and tail() in vector, list, matrix, data.frame, array
#' 
#' @description
#' Concatenates \code{head(n)} and \code{tail(n)} rows and subset with m columns. 
#' Works also with array, list and matrix in data.frame. 
#' Keeps the data.table format (and add a timezone by default). 
#' See the matsindf package for matrix(ces) in tibble.
#' 
#' @param   x      vector, matrix, data.frame, array or list.
#' @param   n      integer. Cut in the first dimension.
#' @param   m      integer. Cut in the second dimension. 
#' @param   p      integer. Cut in the third and the next dimensions. 
#' @param   l      integer. Cut for a list and for data.frame with matrix inside. 
#' @param   names  logical. Provide names and numbers for undefined dimnames.
#' @param   LTT    character. The letter used in each dimension (vector, list, array).
#' @return  
#' An object of the same class than x but much shorter. 
#' 
#' @examples
#' 
#' ### Vector, data.frame, array
#' ht(1:100, names = FALSE)
#' ht(1:100, LTT = "z")
#' 
#' ht9(mtcars); dim(mtcars)
#' 
#' arr4 <- array(1:1680, c(8,7,6,5))
#' ht(arr4, n=1, p=1, names = FALSE)
#' ht(arr4, n=1, p=1, names = TRUE, LTT = c("x","L","X","Y","Z","T"))
#' 
#' ### List of matrices
#' lstmat <- rep(list(matrix(1:100, 10)), 8)
#' for (i in seq_along(lstmat)) lstmat[[i]] <- lstmat[[i]] *i
#' lstmat
#' ht(lstmat, n = 2, m = 2, l = 2, names = FALSE)
#' ht(lstmat, n = 2, m = 3, l = 1, names = TRUE)
#' ht(lstmat, n = 2, m = 3, l = 1, LTT = c("x","L","X","Y","Z"))
#' 
#' ### Data.frame with matrices inside.
#' ### See For instance data("gasoline", package = "pls")
#' 
#' ## Colnames on matrix B but not on matrix C. Protected data.frame.
#' B <- matrix(101:160, 10); colnames(B) <- paste0(1:6, "b"); B
#' C <- matrix(101:160, 10); C
#' dfrmat <- data.frame(A = 1:10, B = I(B), C = I(C), D = 11:20); dfrmat
#' colnames(dfrmat)
#' 
#' ## Matrix columns are controlled by m.
#' ## Unnamed C matrix columns have old values but new names. Be aware!
#' ht(dfrmat, n = 2, m = 1, l = 2, names = FALSE) # Original C.6 is now C.2
#' ht(dfrmat, n = 2, m = 1, l = 2, names = TRUE)  # Names keep original ranks
#' 
#' ## Data.frame columns are controlled by l.
#' ht(dfrmat, n = 3, m = 2, l = 1, names = TRUE)
#' 
#' @export
#' @name ht
ht <- function (x, n = 3, m = 4, p = 2, l = 2, names = TRUE,
                LTT = c("x","L","N","M","P","Q","R","S","T")) {
	fun_ij <- function(i, j) {
		if (j < i/2) c(seq_len(j), seq.int(to=i, length.out=j)) else seq_len(i)
	}
    if (is.null(dim(x))) {
		if (is.list(x)) {
        ## LIST
			if (is.null(names(x)) & names) {
				names(x) <- paste0(LTT[2], seq_along(x))
			}
			xl <- fun_ij(length(x), l)
			x  <- x[xl]
            lapply(x, ht, n = n, m = m, p = p, l = l, names = names, LTT = LTT)
        } else {
		## VECTOR
			if (is.null(names(x)) & names) {
				names(x) <- paste0(LTT[1], seq_along(x))
			}
			xn <- fun_ij(length(x), n)
			x[xn]
        }
    } else {
        if (is.data.frame(x) & ((length(apply(x, 2, length)) != NCOL(x)))) {
		## DATA.FRAME WITH MATRIX INSIDE
			rn  <- rownames(x)
			rnn <- rn[fun_ij(length(rn), n)]
			# message(cat("Matrix in data.frame with colnames", colnames(x)))
			message(paste0("Matrix in data.frame. Colnames are: ", 
			               paste(colnames(x), collapse=", "), 
						   "."))
			data.frame(ht(as.list(x), n=n, m=m, p=p, l=l, names=names, LTT=LTT),
			           row.names = rnn)
        } else {
		## MATRIX, ARRAY AND STANDARD DATA.FRAME. DATA.TABLE since 2020-04-10.
			dd  <- dim(x)
			LT3 <- LTT[-c(1,2)]
			stopifnot(length(dd) <= length(LT3))
			if (names) x <- provideDimnames(x, base = mapply(paste0, 
								as.list(LT3[seq_along(dd)]), 
								lapply(dd, seq_len), 
								SIMPLIFY = FALSE))
			nmp    <- c(n, m, rep(p, length(dd)-2))
			lstdc2 <- mapply(fun_ij, dd, nmp, SIMPLIFY = FALSE)
			listx  <- c(list(x), lstdc2, drop = FALSE)
			if (inherits(x, "data.table")) {
				print(subset(x, select = lstdc2[[2]]), topn = n, timezone = TRUE)
			} else { 
				do.call(`[`, listx)
			}
		}
    }
}

#' @export
#' @rdname ht
ht9 <- function (x, n = 3, m = 9999, p = 2, l = 2, names = TRUE,
                 LTT = c("x","L","N","M","P","Q","R","S","T")) {
	ht(x, n = n, m = m, p = p, l = l, names = names, LTT = LTT)
}





R/a_intro.R
## a_intro = NNbenchmark-package 2019-08-04 + 2019-08-22



#' @title Package NNbenchmark
#' @description
#' Datasets and functions to benchmark (convergence, speed, ease of use) R packages 
#' dedicated to regression with neural networks (no classification in this version).
#' The templates for the tested packages are available at 
#' <https://github.com/pkR-pkR/NNbenchmarkTemplates/tree/master/R> and 
#' <https://akshajverma.com/NNbenchmarkWeb>.
#' @examples
#' ds <- grep("^[m,u]", ls("package:NNbenchmark2019"), value = TRUE); ds
#' t(sapply(ds, function(x) dim(get(x)))) 
#' NNdataSummary(NNdatasets)
#' plot(uGauss2)
#' pairs(mIshigami)
#' ht(NNdatasets, n = 2, l = 6)
#' @import      R6
#' @import      graphics
#' @import      utils
#' @importFrom  stats  sd  formula
#' @importFrom  pkgload  unload
#' @aliases NNbenchmark
#' @name    NNbenchmark-package
#' @docType package
NULL




R/clearNN.R
## clearNN 2019-08-04


#' @title Clean the .GlobalEnv and Detach the Loaded Packages
#' @description
#' \code{clearNN} cleans the first environment \code{ls(.GlobalEnv)} from objects 
#' not listed in \code{donotremove} and detach the loaded packages and their 
#' dependencies not listed in \code{donotdetach}.
#' 
#' \code{detachNN()} removes \code{ZZ} and all objects in \code{ls(.GlobalEnv)} 
#' not listed in \code{donotremove}. It does not detach any packages.
#'  
#' @param  donotremove  a vector of characters representing objects in \code{ls()}.
#' @param  donotdetach  a vector of packages and environments that are not detached.
#'                      \code{donotdetach = NULL} protects the packages and environments 
#'                      \code{".GlobalEnv"}, \code{"package:kableExtra"}, 
#'                      \code{"package:dplyr"}, \code{"package:stringr"}, 
#'                      \code{"package:NNbenchmark"}, \code{"package:rmarkdown"}, 
#'                      \code{"package:knitr"}, \code{"package:captioner"},
#'                      \code{"package:pkgload"}, \code{"package:R6"}, 
#'                      \code{"tools:rstudio"}, 
#'                      \code{"package:RWsearch"}, \code{"package:pacman"}, 
#'                      \code{"package:stats"}, \code{"package:graphics"}, 
#'                      \code{"package:grDevices"}, \code{"package:utils"}, 
#'                      \code{"package:datasets"}, \code{"package:methods"}, 
#'                      \code{"Autoloads"}, \code{"package:base"}.
#' @return  
#' NULL
#' 
#' @export
#' @name clearNN
clearNN <- function (donotremove, donotdetach = NULL) { 
    if (is.null(donotdetach)) donotdetach <- c(".GlobalEnv", 
		   "package:kableExtra", 
           "package:dplyr", "package:stringr", "package:knitr", 
           "package:rmarkdown", "package:captioner", 
		   "package:NNbenchmark", "package:NNbenchmark2019", 
           "package:R6", "package:pkgload", "tools:rstudio", 
           "package:RWsearch", "package:pacman", "package:stats", 
		   "package:graphics", "package:grDevices", "package:utils", 
           "package:datasets", "package:methods", "Autoloads", "package:base")           
    if (!is.element("ZZ", donotdetach)) {    
        while ("ZZ" %in% search()) detach("ZZ", character.only = TRUE)
    }
    if (length(setdiff(search(), donotdetach)) > 0) {
        vecns   <- setdiff(search(), donotdetach)
        strpkgs <- strsplit(vecns, ":", fixed = TRUE)
        lpkgs   <- lengths(strpkgs)
        fun     <- function(l, pkg) {if (l == 2) pkg[2] else NULL}
        pkgs    <- unlist(mapply(fun, lpkgs, strpkgs))
        lapply(pkgs, pkgload::unload, quiet = TRUE)
    }
    if (length(setdiff(search(), donotdetach)) > 0) tryCatch(
           lapply(setdiff(search(), donotdetach), detach, unload = TRUE, 
                  character.only = TRUE, force = TRUE),
           warning = function(w) {})
    # while (length(setdiff(ls(.GlobalEnv), donotremove)) > 0) remove(           
           # list = setdiff(ls(.GlobalEnv), donotremove), envir = .GlobalEnv)
}


#' @export
#' @rdname clearNN
detachNN <- function (donotremove, donotdetach = "") {           
    if (!is.element("ZZ", donotdetach)) {    
        while ("ZZ" %in% search()) detach("ZZ", character.only = TRUE)
    }
    # while (length(setdiff(ls(.GlobalEnv), donotremove)) > 0) remove(           
           # list = setdiff(ls(.GlobalEnv), donotremove), envir = .GlobalEnv)
}



